docs/specs/tsd-006-startup-error-behavior.md

```diff
@@ -0,0 +1,105 @@
+# Startup Error Behavior
+
+## Objective
+
+Unified, configurable behavior on retry-able startup errors.
+
+## Keywords
+
+inputs, outputs, startup, error, retry
+
+## Overview
+
+Many Telegraf plugins connect to an external service either on the same machine
+or via network. On automated startup of Telegraf (e.g. via service) there is no
+guarantee that those services are fully started yet, especially when they reside
+on a remote host. More and more plugins implement mechanisms to retry reaching
+their related service if they failed to do so on startup.
+
+This specification intends to unify the naming of configuration-options, the
+values of those options and their semantic meaning. It describes the behavior
+for the different options on handling startup-errors.
+
+Startup-errors are all errors occurring in calls to `Start()`, for inputs and
+service-inputs, or `Connect()`, for outputs. The behaviors described below
+should only be applied in cases where the plugin *explicitly* states that an
+startup-error is *retry-able*. This comprises for example network errors
+indicating that the host or service is not reachable (yet) or external
+resources, like a machine or file, which are not yet available but might become
+available later. To indicate a retry-able startup-error the plugin should return
+a predefined error-type, potentially including an explicit flag. Additionally,
+the plugin might indicate partial connects to signal
+
+## Configuration Options and Behaviors
+
+Telegraf must introduce a unified `startup_error_behavior` configuration option
+for inputs and output plugins which is handled directly by the Telegraf
+infrastructure and is not passed down to the plugins. The setting must be
+available on a per-plugin basis and defines how Telegraf behaves on startup
+errors.
+
+For all listed behaviors Telegraf might retry to startup the plugin for a
+limited number of times during the startup phase before actually processing
+data.
+
+### `error` behavior
+
+If you set the `error` setting for the `startup_error_behavior` option Telegraf
+must fail and exit on startup-errors. This must be the default behavior.
+
+### `retry` behavior
+
+When using the `retry` setting for the `startup_error_behavior` option Telegraf
+must *not* fail on startup-errors and should continue running. Telegraf must
+retry to startup the failed plugin in each gather- or write-cycle, for inputs
+or for outputs respectively, for an unlimited number of times. Neither the
+plugin's `Gather()` nor `Write()` method is called as long as the startup did
+not succeed.
+
+### `retry-partial` behavior
+
+The `retry-partial` setting for the `startup_error_behavior` option is similar
+to the `retry` setting but the plugin's `Gather()` or `Write()` method is
+called as-soon-as the startup was successful or the returned startup-error
+indicates a partially successful connection, e.g. a subset of the given
+endpoints are reachable.
```

Comment:
Hmm now that I am reading this, I'm wondering why we need to the distinction that we won't call Gather or Write. As a user, I would expect `retry` to keep trying to connect and once it does, start working, even if not all servers are up.

I would propose removing `retry-partial` and collapse this behavior into `retry`. And in retry we need to specify that we keep track of what is connected and only try those servers/endpoints once connected in gather/write.

Reply:
The reason is that on partial connect you want to do both, call `Connect()` to try to connect to the remaining endpoints if possible and `Gather()`/`Write()` to start working. If we remove this case the framework would stop trying to connect as-soon-as the first partial connect happens. See [`outputs.graphite`](https://github.com/influxdata/telegraf/blob/master/plugins/outputs/graphite/graphite.go) for a real-world example!

Reply:
I think my disagreement is with retry's behavior. I don't think we should retry until everything is connected. I think retry should retry until everything is connected and things that are connected should work as expected.

Reply:
Which would break the, explicitly requested, graphite behavior... The changes necessary for also capturing this case are minimal...

Reply:
ok thanks for the discussion yesterday, I do think that `retry-partial` should become the default `retry` option. That appears like a better "default" retry option behavior.

Then the existing `retry` could be `retry-full` or something similar or dropped. I leave that to you, but I think for our first PR we can ignore that option.

Thoughts?